app/controllers/products_controller.rb

```diff
@@ -0,0 +1,63 @@
+class ProductsController < ApplicationController
+  def index
+    @products = Product.all
+    @categories = Category.all
+    @products_array = []
+    @merchant_products = nil
+  end
+
+  def new
+  end
+
+  def category
+    @merchant_products = nil
+    @categories = Category.all
+    @category = Category.find(params[:category])
+    pro = @category.product_categories
+    product_id_array = []
+    pro.each do |item|
+      product_id_array << item.product_id
+    end
+    @products_array = []
+    products = product_id_array.each do |id|
+      @products_array << Product.find(id)
+    end
+
+    render :index
+  end
+
+  def merchant
+    @products_array = []
+    @merchants = User.all
+    @merchant = User.find(params[:user])
+    @merchant_products = @merchant.products
```

Comment:
Since you have access to the `@merchant` instance variable, you can just use this to access the products rather than adding an additional instance variable